Version 1.0 NER4OPT Library

kuppulur opened the pull request:
This is version 1.0 of our library. Please review and let me know any changes you would appreciate. Will address them accordingly.

skadio commented on README.md:
````diff
@@ -1,2 +1,33 @@
+## Installation
+
+```bash
+pip install . 
````
in which folder

skadio commented on README.md:
````diff
@@ -1,2 +1,33 @@
+To run tests, execute:
+
+```
+python -m unittest
````
in which folder

skadio commented on data/README.md:
```diff
@@ -0,0 +1,25 @@
+# Train and Test Data for NER4OPT
+
+Please find the original data repository [here](https://github.com/nl4opt/nl4opt-subtask1-baseline/tree/main/data)
```
Name the challenge.. original NL4OPT @ NeurIPS'22 Data Repository

skadio replied:
BTW, what is the license of this data? Are we allowed to copy here?

skadio replied:
same comment for the codes copied from NL4OPT

skadio replied:
we should probably add a license file to this repo, too?

skadio commented on data/README.md:
````diff
@@ -0,0 +1,25 @@
+
+## Directory structure
+```
+├── train.txt   <- Data containing 713 examples tagged in IOB format
````
add? "Original train data.."

skadio commented on data/README.md:
```diff
@@ -0,0 +1,25 @@
+3. The training set has __713__ examples solely sampled from the Source domain 
+4. The dev has __99__ examples sampled from both Source and Target domain
+
+# To convert the data to json, you can use the following
```
add a word on why is thsi conversation needed and not the original data is not used as-is

skadio commented on models/README.md:
```diff
@@ -0,0 +1,16 @@
+# Models
+
+Please refer [training]() section to understand how our best model is trained
```
refer TO

kuppulur replied:
I added a URL for now, I will re-verify URLs after merge to master.

skadio commented on models/README.md:
```diff
@@ -0,0 +1,16 @@
+2. Lexical Plus Model - In addition to the lexical features, we engineered more optimization specific features followed by a Conditional Random Field model
+3. (Our best) **Hybrid Model** - In addition to lexical features, engineered features, we added transformer predictions as a feature followed by a Conditional Random Field model
+
+## Directory structure
```
the requirements.txt libraries are able to run all these pickles, correct?

kuppulur replied:
If I understand correctly, your question is why we need a bash script right, instead of using requirements.txt file? So, I wasn't able to find an allennlp version, which keeps my torch and spacy version as I wanted. We don't need allennlp for the library, it is a dependency from original NL4Opt Utils. What I did now, which I checked on Ubuntu works, is, install allennlp, overrides first (dependencies for NL4Opt) and then install Ner4Opt. Also, added the updated log.

skadio commented on ner4opt/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+>>> from ner4opt import Ner4Opt
+>>> example_problem = "How much should the firm allocate in each asset so as to maximize its average return ?"
+>>> ner4opt = Ner4Opt('hybrid')  # select a model
```
change Ner4Opt(model="hybrid") and the remove # comment

skadio replied:
In the comment part, you can instead add the possible string options ["lexical", .. "]

skadio commented on ner4opt/__init__.py:
```diff
@@ -0,0 +1,24 @@
+>>> from ner4opt import Ner4Opt
+>>> example_problem = "How much should the firm allocate in each asset so as to maximize its average return ?"
+>>> ner4opt = Ner4Opt('hybrid')  # select a model
+>>> entities = ner4opt.get_entities(example_problem)  # get entities
```
you can drop the # comment.. it is the same as function name

skadio commented on models/README.md:
```diff
@@ -0,0 +1,16 @@
+
+Please refer [training]() section to understand how our best model is trained
+
+There are **three** main models in this repository
```
This can be confusing.. because there are 4 models, including semantic. Can you add a semantic as well and say sth like that is stored in hugging face and link to that store etc.

skadio replied:
That way, explaining Hybrid Model will be even better (since you can now refer to semantic

skadio replied:
you can even add a semantic.txt file here with a note inside that says, the semantic is stored in the hugging face store and add the link

skadio commented on ner4opt/constants.py:
```diff
@@ -0,0 +1,107 @@
+
+    # Deep Models
+    # Roberta Model
+    SEMANTIC_DEEP_MODEL = 'skadio/ner4opt-roberta-v1'
```
To be consistent with the above, why not call it SEMANTIC_MODEL_v1_NAME and SEMANTIC_MODEL_v2_NAME

skadio replied:
OR, even more explicitly.. SEMANTIC_ROBERTA_MODEL

skadio replied:
btw, can we please be explicit about which roberta models are these? Please add their original roberta name/link in the comments here

kuppulur replied:
Yes, I added more details to make things clear about these roberta models

skadio commented on ner4opt/constants.py:
```diff
@@ -0,0 +1,107 @@
+    # Roberta Model
+    SEMANTIC_DEEP_MODEL = 'skadio/ner4opt-roberta-v1'
+    # Roberta Model trained as part of our Hybridization experiment
+    HYBRID_DEEP_MODEL = 'skadio/ner4opt-roberta-v2'
```
I am confused with this naming.. The keyword "hybrid" is not referring to our "Hybrid" model, is it?

skadio replied:
.. hmm .. is this referring to the fine-tuned roberta?? IF SO, consider SEMANTIC_ROBERTA_MODEL and SEMANTIC_ROBERTA_FINE_TUNED_MODEL

kuppulur replied:
I changed naming conventions. The difference in the models is just due to their inherent randomness, the v2 model is the model which got trained as a part of our hybrid modeling experiment, seems to produce slightly better results. V1 model is an older model we trained, we submitted this model to the leaderboard too, but this didn't change our best scores. In the paper, we refer to that as just RoBERTa

skadio commented on ner4opt/ner4opt.py:
```diff
@@ -0,0 +1,135 @@
+class Ner4Opt(object):
+    """Extract named entities for a given optimization problem.
+
+    The main objective of the Ner4Opt library is to generate named entities for optimization problems. There are six
```
Move "There are .. to the next line.. "

Same comment throughout this Text Blob, split sentences/paragraphs over rows.

skadio commented on ner4opt/ner4opt.py:
```diff
@@ -0,0 +1,135 @@
+    OBJ_DIR (objective direction), OBJ_NAME (objective name), PARAM (parameter) and VAR (variable).
+
+    This library provides access to four main methods for extracting these entities.
+    A `lexical` model built using lexical features alone, a `lexical_plus` model which in addition to lexical features
```
Split each model in a new row

skadio commented on ner4opt/ner4opt.py:
```diff
@@ -0,0 +1,135 @@
+    -------
+    text (str) : Sample optimization problem
+
+    Returns named entities of a given optimization problem. A list of dictionaries for each entity
```
should this have Returns______ line?

skadio replied:
is it for each "entity" OR for each "word" ??

kuppulur replied:
I added a returns line. I have looked at different entity output formats last few days. I have to realize that 'O' tags are not useful and generating starting character index , ending character index, score etc for 'O' is wasteful. So, just like standard spacy, huggingface outputs, our output will have "entity phrases" detected and for each we provide, starting/ending character index and a score.

skadio replied:
I like that we are back to standard inputs. Make sure that it is clear in the documentation, pydoc (that we only refer tags for tokens that are predicted to be an entity and omit the rest

skadio commented on ner4opt/ner4opt.py:
```diff
@@ -0,0 +1,135 @@
+
+        self._model_name = model_name
+        self._crf_model = None
+        self._deep_model = None
```
is this the semantic one? consider naming it self._sematinc_model?

kuppulur replied:
I changed the naming conventions and added comments, hope that clarifies.

skadio commented on ner4opt/ner4opt.py:
```diff
@@ -0,0 +1,135 @@
+            self._deep_model = load_torch_model(Constants.SEMANTIC_DEEP_MODEL, use_gpu=use_gpu)
+
+        elif model_name == Constants.HYBRID:
+            self._deep_model = load_torch_model(Constants.HYBRID_DEEP_MODEL, use_gpu=use_gpu)
```
I am probably missing sth.. what's a hybrid deep model? 

I thought our "HYBRID" is using the SEMANTIC_MODEL

skadio commented on notebooks/Data Augmentation.ipynb:
```diff
@@ -0,0 +1,236 @@
+{
```
rename file to data_augmentation.ip

skadio commented on notebooks/Gazetteers.ipynb:
```diff
@@ -0,0 +1,84 @@
+{
```
rename file to gazetteers.ip

skadio commented on notebooks/README.md:
```diff
@@ -0,0 +1,5 @@
+## Directory structure
```
When/why anyone should be using these notebooks??

Can you add a word in this readme about when/why to use these notebooks?

skadio commented on tests/test_ner4opt.py:
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
```
1) not a big fan of matching the output of binary pickle files. How big these pickles would be if we add as dictionary objects inside the tests? OR as files into under tests/data

2) can we add some simple unit tests, with decent problem descriptions against hard-coded output list of dictionaries.. so that we can "see" what is matched. 

3) add a test_usage_example() that is "exactly" the same as the one in readme and assert output

skadio replied:
overall, tests seem quite weak. and ner4opt.utils seem not tested at all. can we add at least a few tests for utilities?

kuppulur replied:
I added more tests and currently the coverage should be close to 99%. I added usage tests for all model types, instead of pickled fixtures, which seem blackboxy, I changed it to json. Reasoning for adding it that way is to make sure, each model output for our known problems is exact and complete, which helps with reproducibility whenever changes are made to main library code. Hope that works. I also added tests for invalid usage, utils and features.

skadio commented on training/README.md:
```diff
@@ -0,0 +1,20 @@
+# Training script for our best model
```
1) why is training folder is under tests folder? how are they related.. 

2) should not the data folder be under training? 
3) should not the notebooks folder be under this training? 

It feels to me, that you can bundle all of this under models folder as in:
models
|_ README.md 
|_ hybrid.pkl
|_ lexical.pkl
|_ lexical_plus.pkl
|_ semantic.txt 
|_training 
|__ data
|__ notebooks
|__ logs
|__ nl4opt_utils
|__ .py and .sh etc.

kuppulur replied:
I don't know why you see tests under training, I don't see it that way. But restructured according to your suggestions.

skadio commented on README.md:
```diff
@@ -1,2 +1,33 @@
+
+Given an optimization problem in natural language, the library extracts **six named entities**:
+
+- **CONST_DIR** (constraint direction)
```
can you use ** CONST_DIR**: Constraint direction and drop the paranthesis thing..

skadio commented on ner4opt/ner4opt.py:
```diff
@@ -0,0 +1,135 @@
+        Name: end, dtype: int: Ending character index of the entity
+        Name: word, dtype: str: Entity phrase
+        Name: entity_group, dtype: str: Type of the entity
+        Name: score, dtype: float: Defines the confidence of the prediction
```
is there a range for this? Is it [0..1]? IF SO, add a comment for its range

skadio commented on ner4opt/ner4opt.py:
```diff
@@ -0,0 +1,135 @@
+        Name: start, dtype: int: Starting character index of the entity
+        Name: end, dtype: int: Ending character index of the entity
+        Name: word, dtype: str: Entity phrase
+        Name: entity_group, dtype: str: Type of the entity
```
Explicitly name the list of groups here.

skadio commented on ner4opt/__init__.py:
```diff
@@ -0,0 +1,24 @@
+**Ner4Opt Library** providing access to the following modules:
+    - ``Ner4Opt``
+
+Example Usage
```
See my changes in the README example and duplicate it here

skadio replied:
BTW, I think we should remove this from here and move it under the ner4opt.py file below the pydoc of Ner4Opt, so taht anyone looking for the code documentation in the main Ner4Opt class will see it

kuppulur replied:
I moved it to the main class.

skadio commented on ner4opt/constants.py:
```diff
@@ -0,0 +1,107 @@
+    # Roberta Model trained as part of our Hybridization experiment
+    HYBRID_DEEP_MODEL = 'skadio/ner4opt-roberta-v2'
+
+    # For generating these keywords automatically, please refer our jupyter notebooks
```
Please add a word on where/how these keywords are used.

skadio commented on ner4opt/constants.py:
```diff
@@ -0,0 +1,107 @@
+        'minimum'
+    ]
+
+    OBJ_DIR_KEYWORDS = [
```
add a comment where this is used

skadio commented on ner4opt/constants.py:
```diff
@@ -0,0 +1,107 @@
+        'overstate', 'most', 'minimize', 'high', 'maximal', 'minimal', 'minimum', 'maximize', 'minimise'
+    ]
+
+    SUBJECTS = ['nsubj', 'nsubjpass', 'csubj', 'csubjpass', 'agent', 'expl']
```
add a comment where this is used

skadio replied:
OR maybe .. if these constants are all used for say Lexical.. Just add a BLOCK COMMENT for constants used in Lexical model add comment blokc #### at the top and bottom of it to keep them separate / distinguished

skadio reviewed (changes requested):
Done with my first pass. Left inline comments; at a high-level

- It looks like some restructuring of the folder might be possible (moving data and notebooks under training, and training under models) 
- some naming inconsistencies of the semantic vs. deep vs.hybrid models in class field names and constants
- opportunity to add a bit more testing (and drop dependency on binary test outputs) 
- polish the usage example 

Overall, GREAT work!

skadio commented on README.md:
````diff
@@ -1,2 +1,33 @@
+
+## Quick Start
+
+```python
````
This usage example IS the heart of all the library here... and it is quire undersold as is.. You need to polish this more.

# Import the Ner4Opt Library
from ner4opt import Ner4Opt

# Problem Description (I am changing example with problem_description, do the same in other places) 
problem_description = "Cautious Asset Investment has a total of $150,000 to manage and decides to invest it in money market fund, which yields a 2% return as well as in foreign bonds, which gives and average rate of return of 10.2%. Internal policies require PAI to diversify the asset allocation so that the minimum investment in money market fund is 40 % of the total investment. Due to the risk of default of foreign countries, no more than 40% of the total investment should be allocated to foreign bonds. How much should the Cautious Asset Investment allocate in each asset so as to maximize its average return?"

# Ner4Opt Model options: lexical, lexical_plus, <add the rest> ..
ner4opt = Ner4Opt(model="hybrid") 

# Extract the list of dictionaries corresponding to each word in the problem description. (this is for each word, if I remember correctly??)
# Each dictionary holds keys for start, end, <add the rest>.. 
entities = ner4opt.get_entities(problem_description)

# Output
print(entities) # if it is better, replace this with how to pretty print it (may be json.print() or prettyprint(), sth like that) 

# Here add the output here and comment out .. if too large output, select a few to give the user a sense of output
# ..
# ..

skadio replied:
# BTW, in the problem description, there are whitespaces before/after single characters (like , % ? etc..) 

I fixed that in your slides as well. I am fixing it in the above too. IS this gonna be a problem? I hope not? Please also fix in other places

kuppulur replied:
The reason for this is, the training, dev, test data are all spacy tokenized. The code works even when the example is not spacy tokenized, but it is better to make sure it is. I added a line about this in data characteristics. It is hard to check for this, we can caution people to usage spacy tokenized example problems as our models are trained on such text. I did mention this in data characteristics, I will also add it at other places, wherever you might think necessary.

skadio commented on README.md:
```diff
@@ -1,2 +1,72 @@
+# Import the Ner4Opt Library
+from ner4opt import Ner4Opt
+
+# Problem Description
```
Here is a good place to add a comment about spacy tokenization.
If you have unspacy free-form text, how would I "spacy-fy" it?

kuppulur replied:
Added a line about spacy tokenization.

skadio replied:
where is that added line? Sorry if i am not seeing sth obvious

skadio replied:
I went up and down a few times but not sure where is this. may be you are referring to some other bullet points under data folder under the training folder.. A typcialy user is unlikely to understand that.. 

My base question is; what happens when we run this without the spaces? Does it not find the same entities? 

Assuming it does not work with "regular" text that is not spacify-ied, we need to tell the user in the readme how to spacify their input. 

Can we add a quick section below the Installation below, sth like "Input Text Format" and explain briefly what is needed and why. Then, can we give the user a one-liner or sth to "spacify" their text (unless we cannot do this automatically under the hood all the time, whenever the user passes the text to us? On that note; what happens if we always "spacify" under the hood to relieve the user from this detail? OR.. it is not gonna work out if we spacify all the time, in case, the text was already spacify-ied)

skadio commented on models/README.md:
```diff
@@ -0,0 +1,23 @@
+
+1. Lexical Model - Uses lexical features like POS tags, Named entities etc., followed by a Conditional Random Field model
+2. Lexical Plus Model - In addition to the lexical features, we engineered more optimization specific features followed by a Conditional Random Field model
+3. Semantic Model - Semantic model is a RoBERTa-based deep learning model fine-tuned on our training data. You find the model [here](https://huggingface.co/skadio/ner4opt-roberta-v1) 
```
"can" find

skadio commented on models/README.md:
```diff
@@ -0,0 +1,23 @@
+1. Lexical Model - Uses lexical features like POS tags, Named entities etc., followed by a Conditional Random Field model
+2. Lexical Plus Model - In addition to the lexical features, we engineered more optimization specific features followed by a Conditional Random Field model
+3. Semantic Model - Semantic model is a RoBERTa-based deep learning model fine-tuned on our training data. You find the model [here](https://huggingface.co/skadio/ner4opt-roberta-v1) 
+4. (Our best) **Hybrid Model** - This model combines the above feature engineering and feature learning models to get best results on our current dataset. 
```
to be explicit, can we add "combines above feature engineering (#2) and "semantic" feature learning (#3) models to get the best performance." You can add a sentence like, see our paper (cite) for more details on our Hybrid Solution. 

You can drop the "on our current dataset". The data is not ours.

skadio replied:
But reading the below.. it sounds like we are not using #3 in #4.. we are using v2? and v2 and v1 are basically same, except some randomness??

kuppulur replied:
I edited to make it more explicit like you suggested. We are not using 3 in 4, for some reason using 3 in 4 gives slightly poorer results even though the model architecture is similar. For the same reasons of uncertainties/randomness, I saved the models we built and submitted to Ner4opt leaderboard too. Regarding citing the paper, do you have any suggestions on where we should upload our paper? Its not publicly available yet right?

skadio commented on models/README.md:
```diff
@@ -0,0 +1,23 @@
+3. Semantic Model - Semantic model is a RoBERTa-based deep learning model fine-tuned on our training data. You find the model [here](https://huggingface.co/skadio/ner4opt-roberta-v1) 
+4. (Our best) **Hybrid Model** - This model combines the above feature engineering and feature learning models to get best results on our current dataset. 
+Please refer to our [training section](https://github.com/skadio/Ner4Opt/training) to understand how our best model is trained. 
+Our RoBERTa based model trained as a part of our hybridization experiment can be found [here](https://huggingface.co/skadio/ner4opt-roberta-v2)
```
It is not clear to the reader what's the difference between this one and the v1 above.

skadio commented on requirements.txt:
```diff
@@ -0,0 +1,11 @@
+inflect
```
Once a user install these requirements, will they be able to run the readme quick start example
```
# Ner4Opt Model options: lexical, lexical_plus, semantic, hybrid
ner4opt = Ner4Opt(model='hybrid')
```
with any model option?

kuppulur replied:
Defaults to hybrid, I updated the readme accordingly.

skadio commented on setup.py:
```diff
@@ -0,0 +1,34 @@
+    install_requires=required,
+    python_requires=">=3.9",
+    classifiers=[
+        "License :: OSI Approved :: MIT License",
```
IF this is gonna be MIT License, let's add that license to the repo

kuppulur replied:
Yes, I will add an MIT license to the repo.

skadio commented on tests/test_utils.py:
```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
```
what's a "fixtures" folder... you mean "data" folder?

kuppulur replied:
One common practice I saw in python testing is to keep the data you need for input/output validation for testing in a folder called fixtures. If you have other suggestions for naming, we can go ahead with that.

skadio replied:
Ok. Interesting. That's new to me. We typically called it simply as data.. that way, the paths to json files look like .."tests/data/input.json" which reads very easily. 

Not important, either way is fine

skadio commented on setup.py:
```diff
@@ -0,0 +1,34 @@
+    exec(fp.read())
+
+setuptools.setup(
+    name="ner4opt",
```
I assume we will push this to PyPI so that people can do `pip install ner4opt`

skadio replied:
yes?

kuppulur replied:
Yes.

skadio commented on models/training/data/README.md:
```diff
@@ -0,0 +1,30 @@
+
+1. The data spans across many domains
+2. The dataset is curated based on two main categories
+   1. __Source__:advertising, investment, sales domains
```
space after ":"

skadio commented on models/training/data/README.md:
```diff
@@ -0,0 +1,30 @@
+1. The data spans across many domains
+2. The dataset is curated based on two main categories
+   1. __Source__:advertising, investment, sales domains
+   2. __Target__:production, science, transportation
```
space after ":"

skadio commented on models/training/data/README.md:
````diff
@@ -0,0 +1,30 @@
+
+## Directory structure
+```
+├── train.txt   <- Original train data containing 713 examples tagged in IOB format
````
I am still unsure, if we can "store" the original data in our own repository. At the top level, it seems we gonna have MIT Licesnse.. Data is subject to another license.. Is this ok?

kuppulur replied:
Great point. I deleted data, added instructions for obtaining the data from the original repository.

skadio commented on models/training/README.md:
```diff
@@ -0,0 +1,33 @@
+pip install allennlp overrides
+
+# Install Ner4Opt
+pip install ner4opt
```
This is gonna come from PyPI i guess?

kuppulur replied:
Yes

skadio commented on models/training/nl4opt_utils/__pycache__/__init__.cpython-39.pyc:
Just checking that __pycache__ is intented to be included here?

kuppulur replied:
Creeped in by mistake, deleted it

skadio commented on models/training/notebooks/README.md:
```diff
@@ -0,0 +1,10 @@
+There are **two** notebooks in this folder
```
very nice!

skadio commented on ner4opt/ner4opt.py:
```diff
@@ -0,0 +1,192 @@
+        Name: start, dtype: int: Starting character index of the entity. Range is [0, len(text)]
+        Name: end, dtype: int: Ending character index of the entity. Range is [0, len(text)]
+        Name: word, dtype: str: Entity phrase
+        Name: entity_group, dtype: str: Type of the entity
```
would be good to spell out all 6 entitiy groups here

kuppulur replied:
Added the groups

skadio commented on ner4opt/ner4opt.py:
```diff
@@ -0,0 +1,192 @@
+    >>> Ner4Opt.__doc__
+    """
+
+    def __init__(self, model: str = Constants.HYBRID, use_gpu: bool = False):
```
The design/naming of _crf_model and _roberta_model has a subtle misleading effect. 
First, they notion of hybrid is now lost, and it is called like roberta model
Second, It ALMOST look like we are not using CRF in the hybrid model which is not true. 

How about the following design: 

```
      # Load model
        if self._model == Constants.LEXICAL:
            self._lexical_model = joblib.load(self._lexical_model_path)

        elif self._model == Constants.LEXICAL_PLUS:
            self._lexical_model = joblib.load(self._lexical_plus_model_path)

        elif self._model == Constants.SEMANTIC:
            self._semantic_model = load_torch_model(Constants.SEMANTIC_MODEL_ROBERTA_V1, use_gpu=use_gpu)

        elif self._model == Constants.HYBRID:
            self._lexical_model = joblib.load(self._hybrid_model_path)
            self._semantic_model = load_torch_model(Constants.SEMANTIC_MODEL_ROBERTA_V2, use_gpu=use_gpu)
```

Notice the replacement of crf_model with lexical_model and roberta_model with semantic model
Notice also the swap of lines under Hybrid when loading to make it consistent

kuppulur replied:
Our (purely) semantic model is the odd one out, which causes the confusion. It relies on the probabilities of entity label predictions and doesn't use CRF as the final layer. All the other models use CRF as the final layer for entity label predictions, with different types of features as inputs (lexical features, lexical features + lexical domain specific features, lexical features +  lexical domain specific features + semantic features). I changed the naming to call all CRF models as _crf_model and deep models as just _semantic_feature_extractor. Also added more comments. Hope that is more clear, please let me know otherwise.

skadio commented on ner4opt/constants.py:
```diff
@@ -0,0 +1,117 @@
+    LEXICAL_PLUS = 'lexical_plus'
+    SEMANTIC = 'semantic'
+    HYBRID = 'hybrid'
+
```
This comment is related design/naming comment in the main function of ner4opt.py. 

These constants give the impression that CRF is not used for deep models. 

How about; 


```
    # Lexical Classical Models 
    LEXICAL_MODEL_NAME = 'lexical.pkl'
    LEXICAL_PLUS_MODEL_NAME = 'lexical_plus.pkl'

    # Semantic Deep Models 
    # https://huggingface.co/skadio/ner4opt-roberta-v1
    # https://huggingface.co/skadio/ner4opt-roberta-v2 (used as part of hybrid)
    # The only difference between v1 and v2 is due inherent randomness during training
    # Refer to the training section for the details of training setup 
    SEMANTIC_MODEL_ROBERTA_V1 = 'skadio/ner4opt-roberta-v1'
    SEMANTIC_MODEL_ROBERTA_V2 = 'skadio/ner4opt-roberta-v2'

    # Hybrid Model using LEXICAL_PLUS_MODEL and SEMANTIC_MODEL_ROBERTA_V2 
    HYBRID_MODEL_NAME = 'hybrid.pkl'
```

kuppulur replied:
Slightly changed it as per your suggestions and also see my commentary above for the new changes.

skadio replied:
this is great, thank you!

skadio reviewed (approved):
Great update over first version. Only minor comments

The only thing is LICENSE, that I am not sure..

kuppulur commented:
> Great update over first version. Only minor comments
> 
> The only thing is LICENSE, that I am not sure..

Thanks. I updated as per your suggestions, removed data to avoid any license conflicts.

skadio commented on tests/test_features.py:
```diff
@@ -0,0 +1,52 @@
+
+class FeaturesTest(unittest.TestCase):
+
+    def setUp(self):
```
very nice set of tests in this file!

skadio commented on tests/test_ner4opt.py:
```diff
@@ -0,0 +1,83 @@
+from ner4opt.exceptions import InvalidModelError
+
+
+class Ner4OptTest(unittest.TestCase):
```
you have a test, test_quick_start() or sth like that, to test the quick start example from the readme, right?

kuppulur replied:
Yes, example in the readme is covered in the test cases.

skadio commented on ner4opt/exceptions.py:
```diff
@@ -0,0 +1,7 @@
+"""Exceptions."""
+
+
+class InvalidModelError(Exception):
```
not important, just curious; what's the advantage of having this vs. raising InvalidValueError()

kuppulur replied:
Thanks for pointing. For now, it does look like an overkill, reverted back to ValueError.

skadio commented on ner4opt/features.py:
```diff
@@ -0,0 +1,592 @@
+import re
```
just noticed one thing; should this file be featurizor.py per the Featurizer() class below?

skadio commented on README.md:
```diff
@@ -1,2 +1,69 @@
+
+## Installation
+
+Ner4Opt requires Python 3.X+ and can be installed from PyPI using `pip install ner4opt` or by building from source 
```
I tweaked this installation section a little and mentioned the PyPI installation

skadio reviewed (approved):
Done. Good to go, no more review from my end. 

Only minor comments that should be easy to care of it and two high-level things in the Readme. 
1) see my note on Installation section
2) comments about Input Text Format. IF we cannot do this under the hood all the time, better we add a section in readme and show the user how to "spacify" their text.

skadio commented on README.md:
```diff
@@ -52,14 +52,37 @@ print("Number of entities found: ", len(entities))
 ## Installation
 
-Ner4Opt requires Python 3.X+ and can be installed from PyPI using `pip install ner4opt` or by building from source 
+Ner4Opt requires Python 3.9+ and can be installed from PyPI using `pip install ner4opt` or by building from source 
```
learning question; why 3.9+ strictly? why not 8 or 7?

kuppulur replied:
Testing with 8 and 7 should work.

kuppulur replied:
Works with 8 and above. I will have to retrain sklearn models for 7 and below.

skadio commented on README.md:
````diff
@@ -1,2 +1,69 @@
+```bash
+git clone https://github.com/skadio/Ner4Opt.git
+cd ner4opt
+pip install .
````
have you looked into this?

kuppulur replied:
Yes. Regarding the wheel file, I will generate and upload to pypi, so a user doesn't have to take that route right? He/she can simply do `pip install .` (similar to `python setup.py install) which should install the package from setup.py which loads the appropriate requirements.txt.

skadio commented on README.md:
````diff
@@ -1,2 +1,92 @@
+Ner4Opt requires Python 3.9+ and can be installed from PyPI using `pip install ner4opt` or by building from source 
+
+```bash
+git clone https://github.com/skadio/Ner4Opt.git
````
fix the name case?

kuppulur replied:
Done
